specifyweb/specify/tree_ranks.py

```diff
@@ -0,0 +1,345 @@
+    'SHELF': 350,
+    'BOX': 400,
+    'RACK': 450,
+    'VIAL': 500
+}
+GEOLOGIC_TIME_PERIOD_RANKS = {
+    'ERA': 100,
+    'PERIOD': 200,
+    'EPOCH': 300,
+    'AGE': 400
+}
+LITHO_STRAT_RANKS = {
+    'SUPERGROUP': 100,
+    'GROUP': 200,
+    'FORMATION': 300,
+    'MEMBER': 400,
+    'BED': 500
+}
+
+DEFAULT_RANK_INCREMENT = 100
+TAXON_RANK_INCREMENT = 10
+GEOGRAPHY_RANK_INCREMENT = DEFAULT_RANK_INCREMENT
+STORAGE_RANK_INCREMENT = 50
+GEOLOGIC_TIME_PERIOD_INCREMENT = DEFAULT_RANK_INCREMENT
+LITHO_STRAT_INCREMENT = DEFAULT_RANK_INCREMENT
+
+# Map tree type to default tree ranks and default rank id increment
+TREE_RANKS_MAPPING = {
+    'taxon': (TAXON_RANKS, TAXON_RANK_INCREMENT),
+    'geography': (GEOGRAPHY_RANKS, GEOGRAPHY_RANK_INCREMENT),
+    'storage': (STORAGE_RANKS, STORAGE_RANK_INCREMENT),
+    'geologictimeperiod': (GEOLOGIC_TIME_PERIOD_RANKS, GEOLOGIC_TIME_PERIOD_INCREMENT),
+    'lithostrat': (LITHO_STRAT_RANKS, LITHO_STRAT_INCREMENT),
+}
+
+TREE_RANK_TO_ITEM_MAP = {
+    'Taxontreedefitem': 'Taxon',
+    'Geographytreedefitem': 'Geography',
+    'Storagetreedefitem': 'Storage',
+    'Geologictimeperiodtreedefitem': 'Geologictimeperiod',
+    'Lithostrattreedefitem': 'Lithostrat'
+}
+
+def get_tree_item_model(tree_rank_model_name):
+    tree_item_model_name = TREE_RANK_TO_ITEM_MAP.get(tree_rank_model_name.title(), None)
+    if not tree_item_model_name:
+        return None
+    return getattr(spmodels, tree_item_model_name, None)
+
+def tree_rank_count(tree_rank_model_name, tree_rank_id) -> int:
+    tree_item_model = get_tree_item_model(tree_rank_model_name)
+    if not tree_item_model:
+        return 0
+    return tree_item_model.objects.filter(definitionitem_id=tree_rank_id).count()
+
+def is_tree_rank_empty(tree_rank_model, tree_rank) -> bool:
+    tree_item_model = get_tree_item_model(tree_rank_model.__name__)
+    if not tree_item_model:
+        return False
+    return tree_item_model.objects.filter(definitionitem=tree_rank).count() == 0
+
+def post_tree_rank_save(tree_def_item_model, new_rank):
+    tree_def = new_rank.treedef
+    parent_rank = new_rank.parent
+    new_rank_id = new_rank.rankid
+
+    # Set the parent rank, that previously pointed to the target, to the new rank
+    child_ranks = tree_def_item_model.objects.filter(treedef=tree_def, parent=parent_rank).exclude(rankid=new_rank_id)
+    if child_ranks.exists():
+        # Iterate through the child ranks, but there should only ever be 0 or 1 child ranks to update
+        for child_rank in child_ranks:
+            child_rank.parent = new_rank
+            child_rank.save()
+
+    # Update the old root rank to point to the new root rank if the new rank is the new root rank
+    if new_rank.parent is None:
+        old_root_rank = tree_def_item_model.objects.exclude(id=new_rank.id).filter(treedef=tree_def, parent=None).first()
+        if old_root_rank is not None:
+            old_root_rank.parent = new_rank
+            old_root_rank.save()
```

Comment:
ha, what use cases does this has? the only case where this branch is taken is if you are adding the same rankid as a previous rank, and the ranks both have the same parent as None. 

instead, simplify things by doing an exclude(id=new_rank.id) instead of exclude(rankid=new_rank_id), above and then getting rid of the lower branch.